frame/support/src/storage/bounded_vec.rs

```diff
@@ -0,0 +1,443 @@
+	pub fn remove(&mut self, index: usize) {
+		self.0.remove(index);
+	}
+
+	/// Exactly the same semantics as [`Vec::swap_remove`].
+	///
+	/// # Panics
+	///
+	/// Panics if `index` is out of bounds.
+	pub fn swap_remove(&mut self, index: usize) {
+		self.0.swap_remove(index);
+	}
+
+	/// Exactly the same semantics as [`Vec::retain`].
+	pub fn retain<F: FnMut(&T) -> bool>(&mut self, f: F) {
+		self.0.retain(f)
+	}
+}
+
+impl<T: BoundedVecValue, S: Get<u32>> TryFrom<Vec<T>> for BoundedVec<T, S> {
+	type Error = ();
+	fn try_from(t: Vec<T>) -> Result<Self, Self::Error> {
+		if t.len() <= Self::bound() {
+			Ok(Self::unchecked_from(t))
+		} else {
+			Err(())
+		}
+	}
+}
+
+// It is okay to give a non-mutable reference of the inner vec to anyone.
+impl<T: BoundedVecValue, S: Get<u32>> AsRef<Vec<T>> for BoundedVec<T, S> {
+	fn as_ref(&self) -> &Vec<T> {
+		&self.0
+	}
+}
+
+// will allow for immutable all operations of `Vec<T>` on `BoundedVec<T>`.
+impl<T: BoundedVecValue, S: Get<u32>> sp_std::ops::Deref for BoundedVec<T, S> {
+	type Target = Vec<T>;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+// Allows for indexing similar to a normal `Vec`. Can panic if out of bound.
+impl<T: BoundedVecValue, S: Get<u32>> sp_std::ops::Index<usize> for BoundedVec<T, S> {
+	type Output = T;
+	fn index(&self, index: usize) -> &Self::Output {
+		self.get(index).expect("index out of bound")
+	}
+}
+
+impl<T: BoundedVecValue, S: Get<u32>> sp_std::iter::IntoIterator for BoundedVec<T, S> {
+	type Item = T;
+	type IntoIter = sp_std::vec::IntoIter<T>;
+	fn into_iter(self) -> Self::IntoIter {
+		self.0.into_iter()
+	}
+}
+
+impl<T: BoundedVecValue, S: Get<u32>> codec::DecodeLength for BoundedVec<T, S> {
+	fn len(self_encoded: &[u8]) -> Result<usize, codec::Error> {
+		// `BoundedVec<T, _>` stored just a `Vec<T>`, thus the length is at the beginning in
+		// `Compact` form, and same implementation as `Vec<T>` can be used.
+		<Vec<T> as codec::DecodeLength>::len(self_encoded)
+	}
+}
+
+impl<T: BoundedVecValue, S: Get<u32>> StorageDecodeLength for BoundedVec<T, S> {}
+
+/// Storage value that is *maybe* capable of [`StorageAppend`].
+pub trait TryAppendValue<T: BoundedVecValue, S: Get<u32>> {
+	/// Try and append the `item` into the storage item.
+	///
+	/// This might fail if bounds are not respected.
+	fn try_append<LikeT: EncodeLike<T>>(item: LikeT) -> Result<(), ()>;
+}
+
+/// Storage map that is *maybe* capable of [`StorageAppend`].
+pub trait TryAppendMap<K: FullCodec, T: BoundedVecValue, S: Get<u32>> {
+	/// Try and append the `item` into the storage map at the given `key`.
+	///
+	/// This might fail if bounds are not respected.
+	fn try_append<LikeK: EncodeLike<K> + Clone, LikeT: EncodeLike<T>>(
+		key: LikeK,
+		item: LikeT,
+	) -> Result<(), ()>;
+}
+
+/// Storage double map that is *maybe* capable of [`StorageAppend`].
+pub trait TryAppendDoubleMap<K1: FullCodec, K2: FullCodec, T: BoundedVecValue, S: Get<u32>> {
```

Comment:
Why do we need 3 versions of this?

```
trait TryStorageAppend<Item> {
    fn try_append(key: impl Fn() -> Vec<u8>, item: Item) -> Result<(), ()>;
}
```

Wouldn't that not work?

Reply:
Not the biggest fan tbh, you can take the same reasoning and refactor a lot of functions from `StorageValue` and `StorageMap` and `StorageDoubleMap` into a common `StorgeBase` using the same method, and while it is less LoC, I think it mostly damaging explicitness and making the code harder to read.

Will give it a try though, maybe I change my mind.

Reply:
One issue would also be that we want to call `decode_len` on all of these before trying to `append`, and double-map's `decode_len` wants the keys individually. Maybe we can refactor `decode_len` to also work with the final key, but I am not interested tbh :D WDYT?

Reply:
We can do:
```rust
trait TryStorageAppend<Item> {
    type Key;
    fn try_append(key: Key, item: Item) -> Result<(), ()>;
}
```
And then use a tuple for map and doublemap, or `()` fo storage value. But I think it makes the API not very good.

Alternatively we can have:

```rust
trait TryStorageAppend<Item> {
    fn try_append(key: Key, item: Item) -> Result<(), ()>;
}
```

We can use the same architecture as traits `StorageAppend` and `StorageDecodeLength`.

We implement `TryStorageAppend` for `BoundedVec`.
For new new macro we can then implement `try_append` for any type which implements `TryStorageAppend` (which is same as what is done in the PR, but more abstract).
For old macros we have do to either those 3 traits, or we add a new method `fn try_append` in `StorageValue`, `StorageMap`, `StorageDoubleMap` with the where clause bounding `TryStorageAppend` similarly to `fn append`.

All in all I think it is ok as it is. But if we want to have other type which would use `try_append` feature, then we need to create this trait.

Reply:
> One issue would also be that we want to call `decode_len` on all of these before trying to `append`, and double-map's `decode_len` wants the keys individually. Maybe we can refactor `decode_len` to also work with the final key, but I am not interested tbh :D WDYT?

https://github.com/paritytech/substrate/blob/a76aadc4b28311418ee114e23b07da46ecccecfb/frame/support/src/storage/mod.rs#L789-L803

It only requires the final key? Not sure what you mean?

Reply:
Right -- yeah, I was looking at the higher level functions like 

https://github.com/paritytech/substrate/blob/a76aadc4b28311418ee114e23b07da46ecccecfb/frame/support/src/storage/mod.rs#L489-L496

Okay so I will try this out.

Reply:
> All in all I think it is ok as it is. But if we want to have other type which would use try_append feature, then we need to create this trait.

I exactly thought of creating that trait and making it look like `StorageAppend` for `Vec<_>`, `DIgest<_>` and a few others. But since there is just one type, it really didn't make sense to me at this point. Maybe a future refactor if we have more bounded types.

Reply:
Okay, so the last issue is that if there is one trait i can't implement it on 3 different generics who have different bounds: 

1. `StorgeValueT: generator::StorageValue`
2. `StorageMapT: generator::StorageMap`
3. etc. 

 Because we'd get conflicting implementations. I am not sure if there is a way around this or not, I tried to follow the pattern of all other storage implementations and implement stuff only for `generator::*` and let the rest be auto-implemented.

Reply:
I will take "silence as consent" here and proceed with merge as I have to move on from this task for now, will be happy to prepare a follow-up if needed.